modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/compaction/Compactor.java

```diff
@@ -164,77 +170,90 @@ void waitDeltaFiles() {
-            }
+            mux.notifyAll();
         }
     }
 
     /**
      * Merges delta files with partition files.
      */
     void doCompaction() {
-        // Let's collect one delta file for each partition.
-        Queue<IgniteBiTuple<FilePageStore, DeltaFilePageStoreIo>> queue = filePageStoreManager.allPageStores().stream()
-                .flatMap(List::stream)
-                .map(filePageStore -> {
+        while (true) {
+            // Let's collect one delta file for each partition.
+            ConcurrentLinkedQueue<DeltaFileToCompaction> queue = new ConcurrentLinkedQueue<>();
+
+            for (GroupPageStores<FilePageStore> groupPageStores : filePageStoreManager.allPageStores()) {
+                for (PartitionPageStore<FilePageStore> partitionPageStore : groupPageStores.getAll()) {
+                    FilePageStore filePageStore = partitionPageStore.pageStore();
+
                     DeltaFilePageStoreIo deltaFileToCompaction = filePageStore.getDeltaFileToCompaction();
 
-                    return deltaFileToCompaction == null ? null : new IgniteBiTuple<>(filePageStore, deltaFileToCompaction);
-                })
-                .filter(Objects::nonNull)
-                .collect(toCollection(ConcurrentLinkedQueue::new));
+                    if (!filePageStore.isMarkedToDestroy() && deltaFileToCompaction != null) {
+                        queue.add(new DeltaFileToCompaction(
+                                groupPageStores.groupId(),
+                                partitionPageStore.partitionId(),
+                                filePageStore,
+                                deltaFileToCompaction
+                        ));
+                    }
+                }
+            }
+
+            if (queue.isEmpty()) {
+                break;
+            }
+
+            updateHeartbeat();
 
-        assert !queue.isEmpty();
+            int threads = threadPoolExecutor == null ? 1 : threadPoolExecutor.getMaximumPoolSize();
 
-        updateHeartbeat();
+            CompletableFuture<?>[] futures = new CompletableFuture[threads];
 
-        int threads = threadPoolExecutor == null ? 1 : threadPoolExecutor.getMaximumPoolSize();
+            for (int i = 0; i < threads; i++) {
+                CompletableFuture<?> future = futures[i] = new CompletableFuture<>();
 
-        CompletableFuture<?>[] futures = new CompletableFuture[threads];
+                Runnable merger = () -> {
+                    DeltaFileToCompaction toMerge;
 
-        for (int i = 0; i < threads; i++) {
-            CompletableFuture<?> future = futures[i] = new CompletableFuture<>();
+                    try {
+                        while ((toMerge = queue.poll()) != null) {
+                            GroupPartitionId partitionId = new GroupPartitionId(toMerge.groupId, toMerge.partitionId);
+
+                            processedPartitionMap.onStartPartitionProcessing(partitionId.getGroupId(), partitionId.getPartitionId());
```

Comment:
Also, it seems that, analogously to `Checkpointer`, all the actual I/O performed by this class must be enclosed in `onStartPartitionProcessing()`/`onFinishPartitionProcessing()` invocations, this is an invariant important for the whole thing to work correctly. If this is true, it would be nice if this requirement was stated explicitly in the class javadoc/comment.

Reply:
For methods `onStartPartitionProcessing` and `onFinishPartitionProcessing` replaced with `GroupPartitionId`